### PR TITLE
[Mem15] Reduce LoopStrider memory usage

### PR DESCRIPTION
**Pull request by andrewcraik**
Currently the loop strider allocates large amounts memory during it recursive analysis routines and this memory leaks during the analysis. This memory leaks until the optimization has finished running which has caused some massive growth in memory consumption. Further, some data structures are pre-sized to the number of symbol references when the actual number of live entries is very small. To reduce this wasted space, several key data structures are transformed from arrays to maps keyed by symref index.

This commit is based on top of Mem 9, 10, 11, 12 and 13 and so should be merged after these other changes. The delta in this commit is 2 commits beyond those in Mem 13.

### CODE REVIEW

**Comment by andrewcraik** [comment]
Rebased onto the merge of Mem 9 - 13 inclusive.

File: compiler/compile/OMRCompilation.cpp [hunk 1]
```diff
@@ -2713,3 +2714,17 @@ OMR::Compilation::CompilationPhaseScope::~CompilationPhaseScope()
    {
    _comp->restoreCompilationPhase(_savedPhase);
    }
+
+TR::Region &OMR::Compilation::aliasRegion()
+   {
+   return self()->_aliasRegion;
+   }
+
```
Comment (vijaysun-omr) [inline compiler/compile/OMRCompilation.cpp]: I would like @lmaisons to also review this commit

File: compiler/optimizer/OMROptimizer.cpp [hunk 2]
```diff
@@ -2050,6 +2050,10 @@ int32_t OMR::Optimizer::performOptimization(const OptimizationStrategy *optimiza
          }
 
       delete opt;
+      // we cannot easily invalidate during IL gen since we could be peeking and we cannot destory our
+      // caller's alias sets
+      if (!isIlGenOpt())
```
Comment (vijaysun-omr) [inline compiler/optimizer/OMROptimizer.cpp]: How does the case when we are doing IL gen for inlining get handled ? i.e. could we be destroying our caller's alias sets (as the comment suggests) in that case ?

File: compiler/compile/OMRCompilation.cpp (hunk 1 above)
Comment (andrewcraik) [inline compiler/compile/OMRCompilation.cpp]: This file is changed in a commit under Mem14 already reviewed by @0xdaryl, but @lmaisons is welcome to comment in Mem14

File: compiler/optimizer/OMROptimizer.cpp (hunk 2 above)
Comment (andrewcraik) [inline compiler/optimizer/OMROptimizer.cpp]: isIlGenOpt will return true in both cases - we found it in the context of peeking but isIlGenOpt will be true for peeking or during inlining. Basically, only after 'true' optimization steps that are not nested can we invalidate the alias set which is the intent behind this check.

Comment (vijaysun-omr) [inline compiler/optimizer/OMROptimizer.cpp]: Okay I see it defers the invalidation until the next "top level compile" opt is finished.

**Comment by vijaysun-omr** [comment]
I'm ready to merge this PR once @lmaisons has looked at the commit I requested he review.

**Comment by andrewcraik** [comment]
Mem14 is WIP so this should not merge until Mem14 has merged.

**Comment by andrewcraik** [comment]
The issue we were tracking has been resolved and this change should be good to merge @vijaysun-omr

**Comment by vijaysun-omr** [comment]
Merging this now that Mem14 has been merged.

**Pull request merged**